### PR TITLE
Add "extended_info" for isolation engine from manifest.

### PR DESCRIPTION
**Pull request by dkarpukhin**

### CODE REVIEW

File: node/include/cocaine/api/isolate.hpp [hunk 1]
```diff
@@ -125,6 +126,11 @@ struct isolate_t {
     spawn(const std::string& path, const args_t& args, const env_t& environment,
                 std::shared_ptr<api::spawn_handle_base_t> handler) = 0;
 
+    virtual
+    std::unique_ptr<cancellation_t>
+    spawn(const std::string& path, const args_t& args, const extended_info_t& extended_info, const env_t& environment,
```
Comment (kletkavrubashku) [inline node/include/cocaine/api/isolate.hpp]: divide line on two

File: node/include/cocaine/service/node/profile.hpp [hunk 2]
```diff
@@ -66,6 +66,7 @@ struct profile_t : cached<dynamic_t> {
     struct {
         std::string type;
         dynamic_t args;
+        dynamic_t extended_info;
```
Comment (kletkavrubashku) [inline node/include/cocaine/service/node/profile.hpp]: is it needed?

File: node/src/isolate/process.cpp [hunk 3]
```diff
@@ -432,6 +432,17 @@ process_t::spawn(const std::string& path,
...
 }
 
+std::unique_ptr<api::cancellation_t>
+process_t::spawn(const std::string& path,
+            const api::args_t& args,
+            const api::extended_info_t& extended_info,
+            const api::env_t& environment,
+            std::shared_ptr<api::spawn_handle_base_t> handle)
+{
+    COCAINE_LOG_WARNING(m_log, "process type cant get benefits for app with extended_info section now");
```
Comment (noxiouz) [inline node/src/isolate/process.cpp]: can't

File: node/src/node/profile.cpp [hunk 4]
```diff
@@ -59,6 +59,7 @@ profile_t::profile_t(context_t& context, const std::string& name_):
 
     isolate.type = isolate_config.at("type", "process").as_string();
     isolate.args = isolate_config.at("args", dynamic_t::empty_object);
+    isolate.args.as_object().insert(std::pair<std::string, dynamic_t>("extended_info", extended_info));
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/src/node/profile.cpp]: fixed

File: node/src/node/app.cpp [hunk 5]
```diff
@@ -657,7 +657,8 @@ app_t::app_t(context_t& context,
     work(std::make_unique<asio::io_service::work>(*loop)),
     thread(nullptr)
 {
-    const profile_t profile{context, profile_name};
+    auto extended_info = manifest_t(context, name).extended_info;
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/src/node/app.cpp]: fixed

File: node/src/isolate/process.cpp [hunk 6]
```diff
@@ -432,6 +432,17 @@ process_t::spawn(const std::string& path,
...
+    process_t::spawn(path, args, environment, handle);
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/src/isolate/process.cpp]: fixed

File: node/include/cocaine/detail/isolate/process.hpp [hunk 7]
```diff
@@ -61,6 +61,10 @@ class process_t:
     spawn(const std::string& path, const api::args_t& args, const api::env_t& environment, std::shared_ptr<api::spawn_handle_base_t>)
         -> std::unique_ptr<api::cancellation_t> override;
 
+    auto
+    spawn(const std::string& path, const api::args_t& args, const api::extended_info_t& extended_info,  const api::env_t& environment, std::shared_ptr<api::spawn_handle_base_t>)
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/include/cocaine/detail/isolate/process.hpp]: fixed

File: node/include/cocaine/detail/isolate/external.hpp [hunk 8]
```diff
@@ -52,6 +52,10 @@ class external_t:
     spawn(const std::string& path, const api::args_t& args, const api::env_t& environment, std::shared_ptr<api::spawn_handle_base_t>)
         -> std::unique_ptr<api::cancellation_t> override;
 
+    auto
+    spawn(const std::string& path, const api::args_t& args, const api::extended_info_t& extended_info, const api::env_t& environment, std::shared_ptr<api::spawn_handle_base_t>)
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/include/cocaine/detail/isolate/external.hpp]: fixed

File: node/include/cocaine/api/isolate.hpp (hunk 1 above)
Comment (dkarpukhin) [reply node/include/cocaine/api/isolate.hpp]: fixed

File: node/include/cocaine/service/node/profile.hpp (hunk 2 above)
Comment (dkarpukhin) [reply node/include/cocaine/service/node/profile.hpp]: not need, test & fixed, tnx

File: node/src/isolate/process.cpp (hunk 3 above)
Comment (dkarpukhin) [reply node/src/isolate/process.cpp]: fixed

File: node/src/node/app.cpp [hunk 9]
```diff
@@ -657,7 +657,9 @@ app_t::app_t(context_t& context,
     work(std::make_unique<asio::io_service::work>(*loop)),
     thread(nullptr)
 {
-    const profile_t profile{context, profile_name};
+    auto manifest = manifest_t(context, name);
+    auto extended_info = manifest.extended_info;
+    const profile_t profile{context, profile_name, extended_info};
```
Comment (kletkavrubashku) [inline node/src/node/app.cpp]: It's extra copying here.
```
    auto manifest = manifest_t(context, name);
    const auto& extended_info = manifest.extended_info;
    const profile_t profile{context, profile_name, extended_info};
```
or
```
    auto manifest = manifest_t(context, name);
    const profile_t profile{context, profile_name,  manifest.extended_info};
```
will be better.

File: node/src/isolate/external.cpp [hunk 10]
```diff
@@ -159,6 +160,22 @@ struct spawn_load_t :
         cancelled(false)
     {}
 
+    spawn_load_t(std::weak_ptr<external_t::inner_t> _inner,
```
Comment (kletkavrubashku) [inline node/src/isolate/external.cpp]: Both constructors are needed?
May be we can use default arg for `extended_info` in causing code.

File: node/src/isolate/external.cpp [hunk 11]
```diff
@@ -542,6 +559,31 @@ external_t::spawn(const std::string& path,
...
 }
 
+std::unique_ptr<api::cancellation_t>
+external_t::spawn(const std::string& path,
+            const api::args_t& worker_args,
+            const api::extended_info_t& extended_info,
+            const api::env_t& environment,
+            std::shared_ptr<api::spawn_handle_base_t> handler) {
+
+    std::shared_ptr<spawn_load_t> load(new spawn_load_t(inner, std::move(handler), path, worker_args, extended_info, environment));
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/src/isolate/external.cpp]: remove code for spawn part, much simplified now

File: node/include/cocaine/api/isolate.hpp [hunk 12]
```diff
@@ -136,6 +142,15 @@ struct isolate_t {
     }
 
 protected:
+    isolate_t(context_t&,
```
[the remark this answers is not in the document]
Comment (dkarpukhin) [reply node/include/cocaine/api/isolate.hpp]: remove code for spawn part, much simplified now

File: node/src/isolate/external.cpp (hunk 10 above)
Comment (dkarpukhin) [reply node/src/isolate/external.cpp]: remove code for spawn part, much simplified now

File: node/src/node/app.cpp (hunk 9 above)
Comment (dkarpukhin) [reply node/src/node/app.cpp]: tnx, fixed

File: node/src/node/app.cpp [hunk 13]
```diff
@@ -670,7 +671,7 @@ app_t::app_t(context_t& context,
 
     state = std::make_shared<app_state_t>(
         context,
-        manifest_t(context, name),
+        manifest,
```
Comment (kletkavrubashku) [inline node/src/node/app.cpp]:
```suggestion
        std::move(manifest),
```

**Review by kletkavrubashku — DISMISSED** [review]

**Comment by kletkavrubashku** [comment]
LGTM, just squash

File: node/src/node/app.cpp [hunk 14]
```diff
@@ -657,7 +657,8 @@ app_t::app_t(context_t& context,
     work(std::make_unique<asio::io_service::work>(*loop)),
     thread(nullptr)
 {
-    const profile_t profile{context, profile_name};
+    auto manifest = manifest_t(context, name);
```
Comment (shaitan) [inline node/src/node/app.cpp]:
```suggestion
    manifest_t manifest{context, name};
```

**Review by shaitan — CHANGES_REQUESTED** [review]

**Review by kletkavrubashku — APPROVED** [review]

**Review by shaitan — APPROVED** [review]

**Pull request merged**